File: client/flutter/divkit/lib/src/core/converters/alignment.dart
import 'package:divkit/divkit.dart';
import 'package:flutter/widgets.dart';

class DivAlignmentConverter {
  final Expression<DivAlignmentVertical?>? vertical;
  final Expression<DivAlignmentHorizontal?>? horizontal;

  const DivAlignmentConverter(this.vertical, this.horizontal);

  AlignmentGeometry? convert() {
    final vertical = this.vertical?.value;
    final horizontal = this.horizontal?.value;

    if (vertical != null && horizontal != null) {
      return vertical.map(
        top: () => horizontal.map(
          left: () => Alignment.topLeft,
          center: () => Alignment.topCenter,
          right: () => Alignment.topRight,
          start: () => AlignmentDirectional.topStart,
          end: () => AlignmentDirectional.topEnd,
        ),
        center: () => horizontal.map(
          left: () => Alignment.centerLeft,
          center: () => Alignment.center,
          right: () => Alignment.centerRight,
          start: () => AlignmentDirectional.centerStart,
          end: () => AlignmentDirectional.centerEnd,
        ),
        bottom: () => horizontal.map(
          left: () => Alignment.bottomLeft,
          center: () => Alignment.bottomCenter,
          right: () => Alignment.bottomRight,
          start: () => AlignmentDirectional.bottomStart,
          end: () => AlignmentDirectional.bottomEnd,
        ),
        // TODO: support baseline alignment in flutter
        baseline: () => horizontal.map(
          left: () => Alignment.centerLeft,
          center: () => Alignment.center,
          right: () => Alignment.centerRight,
          start: () => AlignmentDirectional.centerStart,
          end: () => AlignmentDirectional.centerEnd,
        ),
      );
    } else if (vertical != null) {
      return vertical.map(
        top: () => Alignment.topLeft,
        center: () => Alignment.centerLeft,
        bottom: () => Alignment.bottomLeft,
        // TODO: support baseline alignment in flutter
        baseline: () => Alignment.centerLeft,
      );
    } else if (horizontal != null) {
      return horizontal.map(
        left: () => Alignment.topLeft,
        center: () => Alignment.topCenter,
        right: () => Alignment.topRight,
        start: () => AlignmentDirectional.topStart,
        end: () => AlignmentDirectional.topEnd,
      );
    }
    return null;
  }
}

class DivTextAlignmentConverter {
  final Expression<DivAlignmentVertical> vertical;
  final Expression<DivAlignmentHorizontal> horizontal;

  const DivTextAlignmentConverter(this.vertical, this.horizontal);

  TextAlign convertHorizontal() {
    switch (horizontal.value) {
      case DivAlignmentHorizontal.left:
        return TextAlign.left;
      case DivAlignmentHorizontal.center:
        return TextAlign.center;
      case DivAlignmentHorizontal.right:
        return TextAlign.right;
      case DivAlignmentHorizontal.start:
        return TextAlign.start;
      case DivAlignmentHorizontal.end:
        return TextAlign.end;
    }
  }

  TextAlignVertical convertVertical() {
    switch (vertical.value) {
      case DivAlignmentVertical.top:
        return TextAlignVertical.top;
      case DivAlignmentVertical.center:
        return TextAlignVertical.center;
      case DivAlignmentVertical.bottom:
        return TextAlignVertical.bottom;
      // TODO: support baseline alignment in flutter
      case DivAlignmentVertical.baseline:
        return TextAlignVertical.center;
    }
  }
}

File: client/android/compose/src/main/kotlin/com/yandex/div/compose/utils/Alignment.kt
package com.yandex.div.compose.utils

import androidx.compose.ui.AbsoluteAlignment
import androidx.compose.ui.Alignment
import com.yandex.div2.DivAlignmentHorizontal
import com.yandex.div2.DivAlignmentVertical

internal fun DivAlignmentHorizontal.toHorizontalAlignment(): Alignment.Horizontal =
    when (this) {
        DivAlignmentHorizontal.LEFT -> AbsoluteAlignment.Left
        DivAlignmentHorizontal.RIGHT -> AbsoluteAlignment.Right
        DivAlignmentHorizontal.CENTER -> Alignment.CenterHorizontally
        DivAlignmentHorizontal.END -> Alignment.End
        DivAlignmentHorizontal.START -> Alignment.Start
    }

internal fun DivAlignmentVertical.toVerticalAlignment(): Alignment.Vertical =
    when (this) {
        DivAlignmentVertical.CENTER -> Alignment.CenterVertically
        DivAlignmentVertical.BOTTOM -> Alignment.Bottom
        else -> Alignment.Top
    }

internal fun toAlignment(
    horizontal: DivAlignmentHorizontal,
    vertical: DivAlignmentVertical
): Alignment = combineAlignment(
    horizontal.toHorizontalAlignment(),
    vertical.toVerticalAlignment(),
)

internal fun combineAlignment(
    horizontal: Alignment.Horizontal,
    vertical: Alignment.Vertical,
): Alignment = when (horizontal) {
    AbsoluteAlignment.Left -> when (vertical) {
        Alignment.Top -> AbsoluteAlignment.TopLeft
        Alignment.CenterVertically -> AbsoluteAlignment.CenterLeft
        else -> AbsoluteAlignment.BottomLeft
    }
    AbsoluteAlignment.Right -> when (vertical) {
        Alignment.Top -> AbsoluteAlignment.TopRight
        Alignment.CenterVertically -> AbsoluteAlignment.CenterRight
        else -> AbsoluteAlignment.BottomRight
    }
    Alignment.CenterHorizontally -> when (vertical) {
        Alignment.Top -> Alignment.TopCenter
        Alignment.CenterVertically -> Alignment.Center
        else -> Alignment.BottomCenter
    }
    Alignment.End -> when (vertical) {
        Alignment.Top -> Alignment.TopEnd
        Alignment.CenterVertically -> Alignment.CenterEnd
        else -> Alignment.BottomEnd
    }
    else -> when (vertical) {
        Alignment.Top -> Alignment.TopStart
        Alignment.CenterVertically -> Alignment.CenterStart
        else -> Alignment.BottomStart
    }
}
